mgr/volumes/nfs: move nfs code out of volumes plugin

varshar16 opened the pull request:
This PR is to support integration of dashboard with volumes/nfs plugin.

Fixes: https://tracker.ceph.com/issues/50080
Signed-off-by: Varsha Rao <varao@redhat.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst


## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---
-->

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tchaikov commented:
jenkins test make check

tchaikov commented:
jenkins test api

sebastian-philipp commented on src/pybind/mgr/nfs/export/export_utils.py:
```diff
@@ -0,0 +1,322 @@
+class GaneshaConfParser:
```
do we have unit tests for the conf parser?

varshar16 replied:
> do we have unit tests for the conf parser?

There are no unit tests for it.

batrick replied:
@sebastian-philipp do you have a suggestion on a best-practice location for unit tests that could be written?

epuertat commented on src/pybind/mgr/nfs/export/export_utils.py:
```diff
@@ -0,0 +1,322 @@
+    def stream(self):
+        return self.text[self.pos:]
+
+    def parse_block_name(self):
```
Did you consider reusing some existing pseudo-'parser' like `shlex`? It doesn't fully resolve the parsing but at least it can simplify the process (removes # comments, newlines, whitespace, etc.) and allows you to focus on just processing the `{...}` blocks:
```py
>>> from shlex import shlex

>> config = 'EXPORT{\n    Export_Id = 1 ;   # Export ID unique to each export\n    Path = "volume_path";  # Path of the volume to be exported. Eg: "/test_volume"\n\n    FSAL {\n        name = GLUSTER;\n        hostname = "10.xx.xx.xx";  # IP of one of the nodes in the trusted pool\n        volume = "volume_name";  # Volume name. Eg: "test_volume"\n    }\n\n    Access_type = RW;    # Access permissions\n    Squash = No_root_squash; # To enable/disable root squashing\n    Disable_ACL = TRUE;  # To enable/disable ACL\n    Pseudo = "pseudo_path";  # NFSv4 pseudo path for this export. Eg: "/test_volume_pseudo"\n    Protocols = "3","4" ;    # NFS protocols supported\n    Transports = "UDP","TCP" ; # Transport protocols supported\n    SecType = "sys";     # Security flavors supported\n}\n'

>>> list(shlex.shlex(config, posix=True))
['EXPORT', '{', 'Export_Id', '=', '1', ';', 'Path', '=', 'volume_path', ';', 'FSAL', '{', 'name', '=', 'GLUSTER', ';', 'hostname', '=', '10.xx.xx.xx', ';', 'volume', '=', 'volume_name', ';', '}', 'Access_type', '=', 'RW', ';', 'Squash', '=', 'No_root_squash', ';', 'Disable_ACL', '=', 'TRUE', ';', 'Pseudo', '=', 'pseudo_path', ';', 'Protocols', '=', '3', ',', '4', ';', 'Transports', '=', 'UDP', ',', 'TCP', ';', 'SecType', '=', 'sys', ';', '}']

varshar16 replied:
No, I will look into this in a new PR.

varshar16 commented on src/pybind/mgr/nfs/module.py:
```diff
@@ -0,0 +1,169 @@
+from volumes.module import mgr_cmd_wrap
+import orchestrator
+
+from .export.nfs import NFSCluster, FSExport
```
Not really. I have removed the extra nesting layer.

varshar16 commented on src/pybind/mgr/nfs/module.py:
```diff
@@ -0,0 +1,169 @@
+        except AttributeError:
+            return -errno.EINVAL, "", "Unknown command"
+
+        return handler(inbuf, cmd)
```
How to restrict characters with `@CLICommand` ? Can `goodchars` be passed as argument ?

epuertat replied:
Nope, but given that's for a UUID we can do the following (importing `UUID` from `uuid`):
```diff
- def nfs_export_create_cephfs(self, fsname: str, clusterid: str, binding: str,
+ def nfs_export_create_cephfs(self, fsname: str, clusterid: UUID, binding: str,
```

And in `src/pybind/ceph_argparse.py`, which already has a `CephUUID` type:
```diff
    @staticmethod                                                                                          
    def to_argdesc(tp, attrs, has_default=False):                                                    
        if has_default:  
            attrs['req'] = 'false'    
        CEPH_ARG_TYPES = {                                                                                 
            str: CephString,                                                                               
            int: CephInt,                                                                                  
            float: CephFloat,                                                                              
            bool: CephBool,                                                                                 
+            UUID: CephUUID
        }   
```

varshar16 replied:
```
/home/varsha/ceph/build/bin/ceph -c /home/varsha/ceph/build/ceph.conf nfs cluster create cephfs vstart 
Error EINVAL: Traceback (most recent call last):
  File "/home/varsha/ceph/src/pybind/mgr/mgr_module.py", line 1337, in _handle_command
    return CLICommand.COMMANDS[cmd['prefix']].call(self, cmd, inbuf)
  File "/home/varsha/ceph/src/pybind/mgr/mgr_module.py", line 385, in call
    kwargs = self._collect_args_by_argspec(cmd_dict)
  File "/home/varsha/ceph/src/pybind/mgr/mgr_module.py", line 378, in _collect_args_by_argspec
    kwargs[k] = CephArgtype.cast_to(tp, v)
  File "/home/varsha/ceph/src/pybind/ceph_argparse.py", line 246, in cast_to
    return CephArgtype._cast_to_compound_type(tp, v)
  File "/home/varsha/ceph/src/pybind/ceph_argparse.py", line 231, in _cast_to_compound_type
    raise ValueError(f"unknown type '{tp}': '{v}'  {orig_type} {type_args}")
ValueError: unknown type '<class 'uuid.UUID'>': 'None'
```
It will require more changes to support UUID. I will look into this in a different PR. For now using regex instead.

varshar16 commented on src/pybind/mgr/nfs/module.py:
```diff
@@ -0,0 +1,169 @@
+        return handler(inbuf, cmd)
+
+    @mgr_cmd_wrap
+    def _cmd_nfs_export_create_cephfs(self, inbuf, cmd):
```
Thanks, it is helpful.

varshar16 commented:
jenkins test api

epuertat reviewed (approved):
Apart from the always_on question, everything looks good to me! Thanks @varshar16 !

varshar16 commented on src/mon/MgrMonitor.cc:
```diff
@@ -120,6 +120,7 @@ const static std::map<uint32_t, std::set<std::string>> always_on_modules = {
       "volumes",
       "pg_autoscaler",
       "telemetry",
+      "nfs",
```
Agreed, there is no need for 'nfs' module to be enabled by default. I have updated the PR.

batrick commented on doc/cephfs/fs-nfs-exports.rst:
```diff
@@ -14,6 +14,8 @@ Requirements
 -  ``nfs-ganesha``, ``nfs-ganesha-ceph``, ``nfs-ganesha-rados-grace`` and
    ``nfs-ganesha-rados-urls`` packages (version 3.3 and above)
 
+.. note:: From Quincy nfs mgr module should be enabled
```
```suggestion
.. note:: From Pacific, the nfs mgr module must be enabled prior to use.
```

varshar16 replied:
Done

batrick reviewed (changes requested):
Please explain the reason for moving to a new module in `mgr/volumes/nfs: Move nfs code out of volumes plugin`. In general a title and signed-off-by is insufficient for a commit message.

There are some code maintainability issues that should be addressed (like unit tests) but that need not hold up testing/merging this PR. Please create tracker tickets for these issues.

Rook will now need to enable the nfs module -- I believe?. Please confirm with @leseb. There should be a PendingReleaseNote for this.

Please run through QA after updating to expedite merging.

varshar16 commented:
> Please explain the reason for moving to a new module in `mgr/volumes/nfs: Move nfs code out of volumes plugin`. In general a title and signed-off-by is insufficient for a commit message.
> 

I have updated the commit message and updated the tracker ticket to include link to dashboard integration ticket. Let me know if more detail is required. 

> There are some code maintainability issues that should be addressed (like unit tests) but that need not hold up testing/merging this PR. Please create tracker tickets for these issues.
> 

Here is the tracker ticket: https://tracker.ceph.com/issues/50449

> Rook will now need to enable the nfs module -- I believe?. Please confirm with @leseb. There should be a PendingReleaseNote for this.
> 

Rook does not use the nfs interface. Only the nfs related docs in rook has to be updated.

> Please run through QA after updating to expedite merging.

And I did start testing this PR yesterday but there are some unrelated teuthology failures which is blocking the testing.
` Command failed on smithi171 with status 1: 'TESTDIR=/home/ubuntu/cephtest bash -s' `

https://pulpito.ceph.com/varsha-2021-04-20_13:21:18-rados-wip-reorg-nfs-code-distro-basic-smithi/

batrick commented:
Some of the NFS documentation will need updated as well. Please have a look over that (for example, "cephfs" cluster type is now gone).

varshar16 commented:
> Some of the NFS documentation will need updated as well. Please have a look over that (for example, "cephfs" cluster type is now gone).

The documentation is already updated. (PR https://github.com/ceph/ceph/pull/40411 updated it)

varshar16 commented:
Rebased

varshar16 commented:
@batrick Tests Passed: https://pulpito.ceph.com/varsha-2021-04-22_10:10:28-rados-wip-reorg-nfs-code-distro-basic-smithi/
